api/agent/static_pool.go

```diff
@@ -0,0 +1,146 @@
+package agent
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/fnproject/fn/api/models"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	staticPoolShutdownTimeout = 5 * time.Second
+)
+
+// allow factory to be overridden in tests
+type insecureRunnerFactory func(addr string) (models.Runner, error)
+
+func insecureGRPCRunnerFactory(addr string) (models.Runner, error) {
+	conn, client, err := runnerConnection(addr, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gRPCRunner{
+		address: addr,
+		conn:    conn,
+		client:  client,
+	}, nil
+}
+
+// manages a single set of runners ignoring lb groups
+type staticRunnerPool struct {
+	generator insecureRunnerFactory
+	rMtx      *sync.RWMutex
+	runners   []models.Runner
+}
+
+// DefaultStaticRunnerPool returns a RunnerPool consisting of a static set of runners
+func DefaultStaticRunnerPool(runnerAddresses []string) models.RunnerPool {
+	return newStaticRunnerPool(runnerAddresses, insecureGRPCRunnerFactory)
+}
+
+func newStaticRunnerPool(runnerAddresses []string, runnerFactory insecureRunnerFactory) models.RunnerPool {
+	logrus.WithField("runners", runnerAddresses).Info("Starting static runner pool")
+	var runners []models.Runner
+	for _, addr := range runnerAddresses {
+		r, err := runnerFactory(addr)
+		if err != nil {
+			logrus.WithField("runner_addr", addr).Warn("Invalid runner")
+			continue
+		}
+		logrus.WithField("runner_addr", addr).Debug("Adding runner to pool")
+		runners = append(runners, r)
+	}
+	return &staticRunnerPool{
+		rMtx:      &sync.RWMutex{},
+		runners:   runners,
+		generator: runnerFactory,
+	}
+}
+
+func (np *staticRunnerPool) Runners(call models.RunnerCall) ([]models.Runner, error) {
+	np.rMtx.RLock()
+	defer np.rMtx.RUnlock()
+
+	r := make([]models.Runner, len(np.runners))
+	copy(r, np.runners)
+	return r, nil
+}
+
+func (np *staticRunnerPool) AddRunner(address string) error {
```

Comment:
Minor nit, but what about `rp` rather than `np` for the receiver? (for this method as well as the others in this file)

Reply:
Done